### ISSUE TITLE
Create jinja2 template of ast/ast_common.hpp

### ISSUE THREAD

**Issue by ohm314** [BlueBrain/nmodl#93]
Currently we update ast_common.hpp for every node type (e.g. is_x method for node). This should be done with template. Also, mark those methods const.

**Comment by tapaswenipathak** [comment]
@ohm314 I am interested in this topic, can I work on this ticket?

**Comment by tristan0x** [comment]
There are more than 100 `bool is_SNAKE_NAME()` virtual methods in class `Ast`, where only one of them is overridden in each subclass of Ast. It makes the vtable quite big.

IMO they could be non-virtual methods defined in `Ast`, and the implementations could use the `getNodeType()` method, something like:

```c++
inline bool is_node() const noexcept {
        return getNodeType() == AstNodeType::NODE;
}

inline bool is_statement() const noexcept {
        return getNodeType() == AstNodeType::STATEMENT;
}

// ...
```

Do you concur @ohm314 and @pramodk  ?

**Comment by pramodk** [comment]
> IMO they could be non-virtual methods defined in Ast, and the implementations could use the getNodeType()

Just resounding quickly - 

@tristan0x : the above suggestion will work if you have the pointer of the same class. But in the AST we have many scenarios where the base class points to the derived class pointer and there we need `dynamic dispatch`. e.g.

```c++
Statement *stmt;
ForStatement *for_stmt;
....
stmt = for_stmt;
stmt->is_for_statement(); // need to be dispatched to ForStatement::is_for_statement
```

By the way, there is another similar issue #259. @cattabiani  can provide more info if needed there.

**Comment by cattabiani** [comment]
I remember I discussed this with @pramodk . Following Pramod example: the problem is that a statement is also a forstatement. The first thing that comes into my mind is that we could add a particular function in nodes.py to build all these is_SNAKE_NAME functions only for node such as:

```
inline bool is_statement() const noexcept {
        return getNodeType() == AstNodeType::STATEMENT ||
                   getNodeType() == AstNodeType::FOR_STATEMENT;
}
```

don't pin me down on this implementation. Probably it is not the most performant. It is just for illustration

**Comment by cattabiani** [comment]
the vtable override is less clunky tho

**Comment by pramodk** [comment]
fixed by #290
